Objects/exceptions.c

```diff
@@ -629,6 +630,472 @@ ComplexExtendsException(PyExc_BaseException, SystemExit, SystemExit,
+        matcher->type = EXCEPTION_GROUP_MATCH_BY_TYPE;
+        return 0;
+    }
+    else if (PyFunction_Check(matcher->value)) {
+        matcher->type = EXCEPTION_GROUP_MATCH_BY_PREDICATE;
+        return 0;
+    }
+    PyErr_SetString(
+        PyExc_TypeError,
+        "expected a function, exception type or tuple of exception types");
+    return -1;
+}
+
+static int
+exceptiongroup_split_check_match(PyObject *exc,
+                                 const struct _exceptiongroup_matcher *matcher)
+{
+    switch (matcher->type) {
+    case EXCEPTION_GROUP_MATCH_BY_TYPE: {
+        return PyErr_GivenExceptionMatches(exc, matcher->value);
+    }
+    case EXCEPTION_GROUP_MATCH_BY_PREDICATE: {
+        PyObject *exc_matches = PyObject_CallOneArg(matcher->value, exc);
+        if (exc_matches == NULL) {
+            return -1;
+        }
+        int is_true = PyObject_IsTrue(exc_matches);
+        Py_DECREF(exc_matches);
+        return is_true;
+    }
+    case EXCEPTION_GROUP_MATCH_INSTANCES: {
+        if (PySequence_Check(matcher->value)) {
+            return PySequence_Contains(matcher->value, exc);
+        }
+        return matcher->value == exc;
+    }
+    }
+    return 0;
+}
+
+static PyObject *
+exceptiongroup_split_recursive(PyObject *exc,
+                               const struct _exceptiongroup_matcher* matcher,
+                               int construct_rest)
+{
+    int is_match = exceptiongroup_split_check_match(exc, matcher);
+    if (is_match < 0) {
+        return NULL;
+    }
+
+    if (is_match) {
+        /* Full match */
+        return PyTuple_Pack(2, exc, Py_None);
+    }
+    else if (!PyObject_TypeCheck(exc, (PyTypeObject *)PyExc_BaseExceptionGroup)) {
+        /* Leaf exception and no match */
+        return PyTuple_Pack(
+            2, Py_None, construct_rest ? (PyObject*)exc : Py_None);
+    }
+
+    /* Partial match */
+    PyBaseExceptionGroupObject *eg = _PyBaseExceptionGroupObject_cast(exc);
+    PyObject *match_list = NULL;
+    PyObject *rest_list = NULL;
+    PyObject *match_exc = NULL;
+    PyObject *rest_exc = NULL;
+    PyObject *result = NULL;
+
+    Py_ssize_t num_excs = PySequence_Length(eg->excs);
+    if (num_excs < 0) {
+        goto done;
+    }
+    assert(num_excs > 0); /* checked in constructor, and excs is read-only */
+    match_list = PyList_New(0);
+    if (!match_list) {
+        goto done;
+    }
+    if (construct_rest) {
+        rest_list = PyList_New(0);
+        if (!rest_list) {
+            goto done;
+        }
+    }
+    /* recursive calls */
+    for (Py_ssize_t i = 0; i < num_excs; i++) {
+        PyObject *e = PySequence_GetItem(eg->excs, i);
+        if (!e) {
+            goto done;
+        }
+        PyObject *rec = exceptiongroup_split_recursive(
+            e, matcher, construct_rest);
+        Py_DECREF(e);
+        if (!rec) {
+            goto done;
+        }
+        if (!PyTuple_CheckExact(rec) || PyTuple_GET_SIZE(rec) != 2) {
```

Comment:
This would make more sense as an assert() than a runtime check.
```suggestion
        assert(PyTuple_CheckExact(rec) && PyTuple_GET_SIZE(rec) == 2);
```